### ISSUE TITLE
commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)

### ISSUE THREAD

**Issue by kongove** [scylladb/scylla#2098]
Scylla version (or git commit hash): 1.7.rc0-0.20170207.0429e5d
Cluster size: 6
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS

* Description: *

One db node reported error of allocation when longevity-1.7-gce run for 6+ hours.

* scylla-server log *

scylla-server.service log in longevity-1.7-rc0-gce test: (4.1M)
[longevity-1.7-rc0-scylla-server.log.txt](https://github.com/scylladb/scylla/files/789005/longevity-1.7-rc0-scylla-server.log.txt)

* part of the log* 

```
Feb 20 16:32:24 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 systemd[1]: Starting Scylla Server...
.................
Feb 20 16:32:27 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] database - Populating Keyspace keyspace1
Feb 20 16:32:27 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] database - Keyspace keyspace1: Reading CF standard1
Feb 20 16:32:27 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - Replaying /var/lib/scylla/commitlog/CommitLog-1-12615428.log, /var/lib/scylla/commitlog/CommitLog-1-18014398522097413.log, /var/lib/scylla/commitlog/CommitLog-1-108086391069507333.log, /var/lib/scylla/commitlog/CommitLog-1-108086391069507487.log, /var/lib/scylla/commitlog/CommitLog-1-90071992560025349.log, /var/lib/scylla/commitlog/CommitLog-1-36028797031579397.log, /var/lib/scylla/commitlog/CommitLog-1-54043195541061381.log, /var/lib/scylla/commitlog/CommitLog-1-90071992560025509.log, /var/lib/scylla/commitlog/CommitLog-1-12615429.log, /var/lib/scylla/commitlog/CommitLog-1-36028797031579547.log, 
.................
/var/lib/scylla/commitlog/CommitLog-1-90071992560025507.log, /var/lib/scylla/commitlog/CommitLog-1-126100789578989503.log, /var/lib/scylla/commitlog/CommitLog-1-18014398522097566.log, /var/lib/scylla/commitlog/CommitLog-1-72057594050543551.log
.................
Feb 20 16:35:35 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 5] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-317-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-309-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-301-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-293-Data.db:level=0, ]
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:40 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:41 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] commitlog_replayer - error replaying: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:41 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] database - failed to move memtable for /var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-344-Data.db to cache: std::bad_alloc (std::bad_alloc)
Feb 20 16:35:49 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 5] compaction - Compacted 4 sstables to [/var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-325-Data.db:level=0, ]. 674319072 bytes to 168579768 (~25% of original) in 13744ms = 11.6975MB/s. ~2400256 total partitions merged to 599928.
Feb 20 16:36:06 longevity-1-7-gce-scylla-db-node-d9ea4a62-003 scylla[6796]:  [shard 0] compaction - Compacted 4 sstables to [/var/lib/scylla/data/keyspace1/standard1-915ad4c0f76711e69837000000000003/keyspace1-standard1-ka-448-Data.db:level=0, ]. 456098406 bytes to 160663998 (~35% of original) in 57631ms = 2.65866MB/s. ~1623808 total partitions merged to 571758.
```

**Comment by slivne** [comment]
do we have monitoring for this avilable - if so pling share the info and
ping

**Comment by tgrabiec** [comment]
The node has about 441 segments to replay, and current code replays in parallel. Each reader is using read-ahead of 10, which means 1.2M per reader. That means ~550MB of memory just for buffers (that happens on a single shard).

We should:
  - limit concurrency of replay
  - distribute replay among shards to improve CPU utilization and capacity for concurrency

@kongove What is the amount of memory per instance and per-shard for test?

**Comment by kongove** [comment]
On Tue, Feb 21, 2017 at 5:19 PM, Shlomi Livne <notifications@github.com>
wrote:

> do we have monitoring for this avilable - if so pling share the info and
> ping
>

Oh! I just cleaned the instances.  There is another same job running, I
will ping you and update the PR when the problem is reproduced.

**Comment by kongove** [comment]
6 * db nodes: GCE n1-standard-2 (2 cpu, 7.5G mem)
1 * loader: n1-highcpu-4 (4 cpu, 26G mem)

@tgrabiec

**Comment by slivne** [comment]
replay from a single shard is by desgin - I think I discussed this with @elcallio in the past and we coouldn't move to use multiple shards - @elcallio can you please refer to the relevant issue / provide the reason.

**Comment by kongove** [comment]
Longevity test in AWS:
6 * db nodes: i2.4xlarge (16 cpu, 122G mem)
1 * loader: c3.large (2 cpu, 3.75 mem)

@slivne @roydahan do we need to use large instances for longevity-gce.
```
db node: n1-highmem-16 (16 cpu, 104G mem)
loader: n1-standard-2 (2 cpu, 7.5G mem)
```

**Comment by tgrabiec** [comment]
It seems that segments are already processed in parallel, recover() is using `map_reduce` on files. Couldn't we process those files in parallel on all shards instead of on one shard?

What kind of reordering are you talking about? mutations in commitlog segments are not in ring order, in general.

**Comment by elcallio** [comment]
On 2017-02-21 11:32, Tomasz Grabiec wrote:
>
> It seems that segments are already processed in parallel, recover() is 
> using |map_reduce| on files. Couldn't we process those files in 
> parallel on all shards instead of on one shard?
>
That is indeed true (was not initially me thinks). But its a good 
argument that it could just as well be run on multiple cpus.
>
> What kind of reordering are you talking about? mutations in commitlog 
> segments are not in ring order, in general.
>
No, but segments are in "apply-order", i.e. the same replay position 
ordering required by normal mutation apply (which however of course is 
only enforced on memtable flush boundaries) is in the mutation order. 
But as I said, I doubt it has any relevant benefits (or downsides) to 
try to uphold this ordering of re-applying the commit log mutations. 
Origin of course does, but I'm guessing its more because of "thats how 
it is".
And of couse, as stated above, we don't really uphold this now anyway.

**Comment by tgrabiec** [comment]
2017-02-21 12:56 GMT+01:00 Calle Wilund <notifications@github.com>:

>
>
> On 2017-02-21 11:32, Tomasz Grabiec wrote:
> >
> > It seems that segments are already processed in parallel, recover() is
> > using |map_reduce| on files. Couldn't we process those files in
> > parallel on all shards instead of on one shard?
> >
> That is indeed true (was not initially me thinks). But its a good
> argument that it could just as well be run on multiple cpus.
> >
> > What kind of reordering are you talking about? mutations in commitlog
> > segments are not in ring order, in general.
> >
> No, but segments are in "apply-order", i.e. the same replay position
> ordering required by normal mutation apply (which however of course is
> only enforced on memtable flush boundaries) is in the mutation order.

But as I said, I doubt it has any relevant benefits (or downsides) to
> try to uphold this ordering of re-applying the commit log mutations.
> Origin of course does, but I'm guessing its more because of "thats how
> it is".
> And of couse, as stated above, we don't really uphold this now anyway.
>

If we require replay positions to be monotonic in apply(), is it then
correct to replay out of order?

Memtable flush can happen anywhere during replay. If we crash after
flushing memtables with max position before the position of not-yet-applied
mutations from other segments, could we incorrectly consider those
not-yet-replayed segments as already replayed during next boot?

**Comment by elcallio** [comment]
On 2017-02-21 12:35, Tomasz Grabiec wrote:
> 2017-02-21 12:56 GMT+01:00 Calle Wilund <notifications@github.com>:
>
> >
> >
> > On 2017-02-21 11:32, Tomasz Grabiec wrote:
> > >
> > > It seems that segments are already processed in parallel, recover() is
> > > using |map_reduce| on files. Couldn't we process those files in
> > > parallel on all shards instead of on one shard?
> > >
> > That is indeed true (was not initially me thinks). But its a good
> > argument that it could just as well be run on multiple cpus.
> > >
> > > What kind of reordering are you talking about? mutations in commitlog
> > > segments are not in ring order, in general.
> > >
> > No, but segments are in "apply-order", i.e. the same replay position
> > ordering required by normal mutation apply (which however of course is
> > only enforced on memtable flush boundaries) is in the mutation order.
>
> But as I said, I doubt it has any relevant benefits (or downsides) to
> > try to uphold this ordering of re-applying the commit log mutations.
> > Origin of course does, but I'm guessing its more because of "thats how
> > it is".
> > And of couse, as stated above, we don't really uphold this now anyway.
> >
>
> If we require replay positions to be monotonic in apply(), is it then
> correct to replay out of order?
>
They need only be monotonic in regards to the high RP recorded on last 
flush.

> Memtable flush can happen anywhere during replay. If we crash after
> flushing memtables with max position before the position of 
> not-yet-applied
> mutations from other segments, could we incorrectly consider those
> not-yet-replayed segments as already replayed during next boot?
Again, two issue discussions are overlapping. The memtable::applys in 
replay do not record a
replay position, so any sstable writes caused by the replay will have 
zero RP:s, and
will thus force full replay if considered in finding new low watermarks.

This is behaviour inherited. For us, we could not even try to use old 
RP:s in memtables generated by replay, since
the old segments might be from an application run with a cpu set 
different from now, i.e. sharding will be off.

**Comment by tgrabiec** [comment]
2017-02-21 14:02 GMT+01:00 Calle Wilund <notifications@github.com>:

>
>
> On 2017-02-21 12:35, Tomasz Grabiec wrote:
> > 2017-02-21 12:56 GMT+01:00 Calle Wilund <notifications@github.com>:
> >
> > >
> > >
> > > On 2017-02-21 11:32, Tomasz Grabiec wrote:
> > > >
> > > > It seems that segments are already processed in parallel, recover()
> is
> > > > using |map_reduce| on files. Couldn't we process those files in
> > > > parallel on all shards instead of on one shard?
> > > >
> > > That is indeed true (was not initially me thinks). But its a good
> > > argument that it could just as well be run on multiple cpus.
> > > >
> > > > What kind of reordering are you talking about? mutations in commitlog
> > > > segments are not in ring order, in general.
> > > >
> > > No, but segments are in "apply-order", i.e. the same replay position
> > > ordering required by normal mutation apply (which however of course is
> > > only enforced on memtable flush boundaries) is in the mutation order.
> >
> > But as I said, I doubt it has any relevant benefits (or downsides) to
> > > try to uphold this ordering of re-applying the commit log mutations.
> > > Origin of course does, but I'm guessing its more because of "thats how
> > > it is".
> > > And of couse, as stated above, we don't really uphold this now anyway.
> > >
> >
> > If we require replay positions to be monotonic in apply(), is it then
> > correct to replay out of order?
> >
> They need only be monotonic in regards to the high RP recorded on last
> flush.
>
> > Memtable flush can happen anywhere during replay. If we crash after
> > flushing memtables with max position before the position of
> > not-yet-applied
> > mutations from other segments, could we incorrectly consider those
> > not-yet-replayed segments as already replayed during next boot?
> Again, two issue discussions are overlapping. The memtable::applys in
> replay do not record a
> replay position, so any sstable writes caused by the replay will have
> zero RP:s, and
> will thus force full replay if considered in finding new low watermarks.
>
> This is behaviour inherited. For us, we could not even try to use old
> RP:s in memtables generated by replay, since
> the old segments might be from an application run with a cpu set
> different from now, i.e. sharding will be off.
>

I see.

**Comment by slivne** [comment]
One such "bad case" would be if commitlog files are removed and the node is
booted up without replaying the "rest" of the commitlog

How does this work with schema changes - especially when a column is
dropped and then added back, I reclal vagely we had schema
mutations/markers embedded as well - and its not clear to me if it was ok
to replay out of order with regards to those.

**Comment by elcallio** [comment]
Schema versions per segment, so reordering the segments should be ok.

**Comment by avikivity** [comment]
Given read-ahead, do we really need per-shard concurrency? As long as we allow memtable applies to run in parallel, I think we can replay one segment/shard.

We can do something like this:

 1. each shard replays the segments that ought to belong to it if no resharding occurs; so in the common case everything is local
 2. each shard replays one segment at a time
 3. each segment replays many mutations concurrently, in case apply has to go to a different shard and wait for it

**Comment by tgrabiec** [comment]
Sounds fine to me.

Read-ahead doesn't work across files, so the CPU would become idle when moving from one segment to another, waiting for the first buffer of the next segment. I would expect it to reduce throughput by a small fraction given that segments are large. If that should be significant, we could have a pipeline with two fibers, segment data reader and replayer.

**Comment by elcallio** [comment]
Smallest fix would be to change replay to be single shard sequential on files, parallel on mutations. 

Shard segments will improve thoughput (not sure how much of an issue that is). 

We might also want to add a throttle to mutation generation as well, since we could still potentially generate quite a lot from even a single segment. Though waiting for mutations from each segment to "finish" before starting the next should be more than enough probably.

**Comment by avikivity** [comment]
A concurrency of 2  segments per shard would allow one segment to continue supplying mutations while the other is being closed, so we don't have to sacrifice throughput.

**Comment by slivne** [comment]
replay of commitlog is something we need todo when we recover from a failed node after a crash - is there a real reason to try and optimize this to be as efficient as possible - wouldn't it be better to have a simpler solution ?

**Comment by avikivity** [comment]
It should be efficient, on a large node we can have half a terabyte in commitlog, we don't want to wait an hour for it to replay.

That said, simply replaying per shard would bring an enormous improvement by itself.  Even if we replay a single segment per shard, any pause will be used by the memtable writers and maybe compaction, so it's not that bad.

**Comment by ibalashov** [comment]
Could anyone suggest a (temporary maybe) workaround for this issue?

We do have a cluster of 5 nodes n2-standard 7.5Gb in GCE (cluster load about 0.5TB per node, with 2TB disks per node to reach high IOPS), after several days one node started to fail with the above error. The worst part is that clients are not able to proceed, failing with `com.datastax.driver.core.exceptions.UnavailableException: Not enough replica available for query at consistency ONE (1 required but only 0 alive)`.

**Comment by tgrabiec** [comment]
@ibalashov What's your replication factor? CL=1 query should not fail with one node down unless your replication factor is 1.

What error message do you see exactly? Does it happen after you started a node? This issue is about error during commitlog replay, which is done when a node starts, but it is not clear from your message that this is what happens, since you write "after several days one node started to fail ...".

**Comment by ibalashov** [comment]
@tgrabiec Replication factor is 2 (set on keyspace).

After the initial failure the node entered crash/restart loop, possibly due to additional flags `--abort-on-lsa-bad-alloc=1 --abort-on-seastar-bad-alloc`.

Here is an extract from the log:
https://gist.github.com/ibalashov/a7961197a24a7cb991012fd99540e8e5

Just now I made an attempt to realive the node by deleting the commit log files, which didn't work, and the error is different now for obvious reasons, but that's another story.

**Comment by tgrabiec** [comment]
@ibalashov Do you have a core dump for the initial error? What's the new error?

**Comment by ibalashov** [comment]
> Do you have a core dump for the initial error?

Unfortunately, no.

```
root@scylla-gce-us-2:~# ls -la /var/lib/scylla/coredump/
total 8
drwxr-xr-x 2 scylla scylla 4096 Feb 18 06:48 .
drwxr-xr-x 5 scylla scylla 4096 Mar  1 07:24 ..

root@scylla-gce-us-2:~# coredumpctl info
No coredumps found.
```

> What's the new error?

The error was something about commit log files missing, aborting. 

Anyway, after `nodetool removenode foo ; nodetool removenode force` the clients are back to working state at least (at the expense of one lost node, but I can wipe it up and re-add to cluster I guess).

**Comment by tgrabiec** [comment]
@ibalashov It would be good to know why coredumps are not generated, they may be critical to solving some problems. Is there anything in system logs which gives a hint why it was not generated? Maybe not enough disk space?

**Comment by ibalashov** [comment]
@tgrabiec Here is what I see in syslog:

> Mar 31 11:41:52 scylla-gce-us-2 systemd-coredump[18731]: Failed to send coredump datagram: Message too long
Mar 31 11:41:52 scylla-gce-us-2 systemd[1]: Started Process Core Dump (PID 18731/UID 0).
Mar 31 11:41:52 scylla-gce-us-2 systemd-coredump[18732]: Coredump file descriptor missing.
Mar 31 11:41:52 scylla-gce-us-2 systemd[1]: scylla-server.service: Failed with result 'core-dump'.

Here are some more system info, jic:
https://gist.github.com/ibalashov/62ac8c41ab391b6c7cb4051107ece12b
